remove CompleteTensorType

zdevito opened the pull request:
This replaces uses of CompleteTensorType with ProfiledTensorType. 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24169 remove CompleteTensorType**
* #24077 Remove DimensionedTensorType

Differential Revision: [D16765329](https://our.internmc.facebook.com/intern/diff/D16765329)

zdevito committed:
remove CompleteTensorType

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

Krovatkin commented on torch/csrc/jit/fuser/tensor_desc.h:
```diff
@@ -41,8 +41,11 @@ struct TORCH_API TensorDesc {
-  TensorDesc(const c10::CompleteTensorTypePtr& type)
-      : TensorDesc(type->scalarType(), type->sizes(), type->strides()) {}
+  TensorDesc(const c10::ProfiledTensorTypePtr& type)
+      : TensorDesc(
```
nitpick: probably won't hurt to stick in `INTERNAL_TORCH_ASSERT(type->isComplete())` in places like this

Krovatkin commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -658,6 +622,10 @@ struct CAFFE2_API ProfiledTensorType : public TensorType {
     auto gr = merge_primitive(requiresGrad(), other->requiresGrad());
     return ProfiledTensorType::create(scalar_type, dev, sz, srs, gr);
   }
+  // is all information about the type specified except for autograd?
```
maybe we could a quick comment as to why `requires_gard` is excluded?

Krovatkin commented on torch/csrc/jit/passes/peephole.cpp:
```diff
@@ -46,9 +46,11 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
         auto expanded_sizes = node->get<c10::List<int64_t>>(attr::size);
-        if (!expanded_sizes.has_value() || c10::impl::toVector(*expanded_sizes) == input_type->sizes()) {
+        auto input_type_sizes = input_type->sizes().concrete_sizes();
+        if (expanded_sizes.has_value() && input_type_sizes &&
```
is omitting `!` in front of `expanded_sizes.has_value()` intentional?

zdevito replied:
yes, my reading was that this code was completely bogus before this change. Before it said 'if expanded_sizes is not a constant, then it is safe to remove the expand call'. Now it says 'if expanded_sizes is a constant, and  its value is the same as the input sizes, we can remove the expand'.

Krovatkin reviewed (approved):
:shipit: a few minor comments

zdevito commented on test/test_jit.py:
```diff
@@ -8209,9 +8209,9 @@ def tensor_unifying(x, y, z):
-        self.assertTrue(if_outputs[0].type().str() == "Float(2, 2)")
-        self.assertTrue(if_outputs[1].type().str() == "Tensor(2, *)")
-        self.assertTrue(if_outputs[2].type().str() == "Tensor(2, 4)")
+        self.assertTrue(if_outputs[0].type().str() == "Float(*, *)")
```
I changed this back to what it was before 'remove DimensionedTensorType', now propagate shapes only sets the dimensions and not the sizes. There were some tests that relied on this behavior.

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head

zdevito committed:
Update on "remove CompleteTensorType"

remove CompleteTensorType

gh-metadata: pytorch pytorch 24169 gh/zdevito/83/head
